app/src/main/java/dev/arkbuilders/arkmemo/graphics/SVG.kt

```diff
@@ -0,0 +1,215 @@
+package dev.arkbuilders.arkmemo.graphics
+
+import android.graphics.Color
+import android.graphics.Paint
+import android.graphics.Path as AndroidDrawPath
+import android.util.Xml
+import dev.arkbuilders.arkmemo.ui.viewmodels.DrawPath
+import org.xmlpull.v1.XmlPullParser
+import java.nio.file.Path
+import java.util.Stack
+import kotlin.io.path.reader
+import kotlin.io.path.writer
+
+class SVG {
+    private var strokeColor = "black"
+    private var fill = "none"
+    private var viewBox = "0 0 100 100"
+    private val commandsArray = ArrayDeque<SVGCommand>()
+    private val canvasPaths = Stack<DrawPath>()
+
+    private val paint
+        get() = Paint().also {
+            it.color = Color.parseColor(strokeColor)
+            it.style = Paint.Style.STROKE
+            it.strokeWidth = 10f
+            it.strokeCap = Paint.Cap.ROUND
+            it.strokeJoin = Paint.Join.ROUND
+            it.isAntiAlias = true
+        }
+
+    fun addCommand(command: SVGCommand) {
+        commandsArray.add(command)
+    }
+
+    fun addPath(path: DrawPath) {
+        canvasPaths.add(path)
+    }
+
+    fun setViewBox(width: Float, height: Float) {
+        viewBox = "0 0 $width $height"
+    }
+
+    fun generate(path: Path) {
+        if (commandsArray.isNotEmpty()) {
+            val xmlSerializer = Xml.newSerializer()
+            val pathData = commandsArray.joinToString(COMMA)
+            xmlSerializer.apply {
+                setOutput(path.writer())
+                startDocument("utf-8", false)
+                startTag("", SVG_TAG)
+                attribute("", Attributes.VIEW_BOX, viewBox)
+                attribute("", Attributes.XML_NS_URI, XML_NS_URI)
+                startTag("", PATH_TAG)
+                attribute("", Attributes.Path.STROKE, strokeColor)
+                attribute("", Attributes.Path.FILL, fill)
+                attribute("", Attributes.Path.DATA, pathData)
+                endTag("", PATH_TAG)
+                endTag("", SVG_TAG)
+                endDocument()
+            }
+        }
+    }
+
+    fun getPaths() = canvasPaths
```

Comment:
Provide type annotation for public methods returning values. It could be a good idea to generalize type, e.g. return `Iterable` instead of specific types like `Stack` or `ArrayDeque`. Double-check the order of iteration when you generalize types.

Reply:
Order of iteration should be from the first element to the last
Do we really have a good reason to generalize types? I see ```Iterable<>``` doesn't have ```isNotEmpty```, can't we have ```IndexOutOfBounds``` error there?